File: patterns/go/top_k_elements.go
package main

import (
	"container/heap"
	"sort"
)

// ********** K Largest Elements **********
// K Largest Elements using Sorting
func kLargestElementsSortingApproach(nums []int, k int) []int {
	sort.Sort(sort.Reverse(sort.IntSlice(nums)))
	return nums[:k]
}

// K Largest Elements using Max Heap
func kLargestElementsMaxHeapApproach(nums []int, k int) []int {
	h := &MaxHeap{}
	heap.Init(h)
	for _, num := range nums {
		heap.Push(h, num)
	}
	result := make([]int, k)
	for i := 0; i < k; i++ {
		result[i] = heap.Pop(h).(int)
	}
	return result
}

// K Largest Elements using Min Heap
func kLargestElementsMinHeapApproach(nums []int, k int) []int {
	h := &MinHeap{}
	heap.Init(h)
	for i := 0; i < k; i++ {
		heap.Push(h, nums[i])
	}
	for i := k; i < len(nums); i++ {
		heap.Push(h, nums[i])
		if h.Len() > k {
			heap.Pop(h)
		}
	}
	result := make([]int, k)
	for i := 0; i < k; i++ {
		result[i] = heap.Pop(h).(int)
	}
	return result
}

// ********** Helper Structures **********

type MaxHeap []int
type MinHeap []int

func (h MaxHeap) Len() int            { return len(h) }
func (h MaxHeap) Less(i, j int) bool  { return h[i] > h[j] } // Max heap
func (h MaxHeap) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *MaxHeap) Push(x interface{}) { *h = append(*h, x.(int)) }
func (h *MaxHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

func (h MinHeap) Len() int            { return len(h) }
func (h MinHeap) Less(i, j int) bool  { return h[i] < h[j] } // Min heap
func (h MinHeap) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *MinHeap) Push(x interface{}) { *h = append(*h, x.(int)) }
func (h *MinHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

// ********** Main **********

func main() {
	// Example test cases
	nums := []int{3, 2, 1, 5, 6, 4}
	k := 2

	// Sorting Approach
	result := kLargestElementsSortingApproach(nums, k)
	fmt.Println("K Largest Elements (Sorting Approach):", result)

	// Max Heap Approach
	result = kLargestElementsMaxHeapApproach(nums, k)
	fmt.Println("K Largest Elements (Max Heap Approach):", result)

	// Min Heap Approach
	result = kLargestElementsMinHeapApproach(nums, k)
	fmt.Println("K Largest Elements (Min Heap Approach):", result)
}

File: patterns/python/top_k_elements.py
import heapq
from collections import Counter

class TopKElements:
    
    # K Largest Elements using Sorting
    def k_largest_elements_sorting_approach(self, nums, k):
        return sorted(nums, reverse=True)[:k]

    # K Largest Elements using Max Heap
    def k_largest_elements_max_heap_approach(self, nums, k):
        return heapq.nlargest(k, nums)

    # K Largest Elements using Min Heap
    def k_largest_elements_min_heap_approach(self, nums, k):
        min_heap = nums[:k]
        heapq.heapify(min_heap)
        for num in nums[k:]:
            heapq.heappush(min_heap, num)
            if len(min_heap) > k:
                heapq.heappop(min_heap)
        return [heapq.heappop(min_heap) for _ in range(k)][::-1]

    # Top K Frequent Elements using Sorting
    def top_k_frequent_elements_sorting_approach(self, nums, k):
        count = Counter(nums)
        return [num for num, freq in count.most_common(k)]

    # Top K Frequent Elements using Min Heap
    def top_k_frequent_elements_min_heap_approach(self, nums, k):
        count = Counter(nums)
        min_heap = []
        for num, freq in count.items():
            heapq.heappush(min_heap, (freq, num))
            if len(min_heap) > k:
                heapq.heappop(min_heap)
        return [heapq.heappop(min_heap)[1] for _ in range(k)][::-1]

    # K Closest Points to Origin using Max Heap
    def get_distance(self, point):
        return point[0] ** 2 + point[1] ** 2

    def k_closest_points_to_origin_max_heap_approach(self, points, k):
        max_heap = []
        for point in points:
            heapq.heappush(max_heap, (-self.get_distance(point), point))
            if len(max_heap) > k:
                heapq.heappop(max_heap)
        return [heapq.heappop(max_heap)[1] for _ in range(k)][::-1]